### DISCUSSION TITLE
SIM7600 library for micropython

### DISCUSSION THREAD

**Discussion by andypnz** [micropython/micropython#11649] [Libraries & Drivers] [answered]
Hi. I am using MicroPython on a RP2040 and have a SIM7600 board attached.

I have an existing project which monitors some sensors and sets off a buzzer, etc if a sensor is out of range. The project uses asyncio.

I would like to add sending and receiving SMS messages to this project.

Is there a library available that will let me do that?

Thank you!

**Comment by 2dof** [comment]
use  AT commands via UART  between rp2040 and module.  All AT command you have in product description, so from RP side you have to 
init UART connection and prepare At commands  and send them. 
for uart use asyncio  StreamReader/StreamWriter  to  commands to port ( [see uasyncio and SIM800 UART](https://forum.micropython.org/viewtopic.php?t=4603)

**Comment by andypnz** [comment]
Hi... Thank you for the link and suggestion :-) I am using the class SIM_UART code as a class, and calling it from my main file async def sms_loop():

The first problem I face is I can't seem to find a link to the pyb library (from pyb import UART) - so instead I am using (from machine import UART, Pin) - and this is I think working - and this lets me define the Tx/Rx pins. I am using a RP2040 board.

And secondly (and I am sure I am doing something quite wrong here) I need to send the AT command twice to send a SMS - and the format is not as I expected - but it does send a SMS :-)

The first time I send res = await sim_uart.send_command('AT+CMGS="+6421123456"\rHello User\x1A') it does not send a SMS, and I see in the log (the ? in black here is actually a square box in the log...):
```
INFO:__main__:sim.command...
Hello User+6421123456"
Result 2 is: b'\xffAT+CMGS="+6421123456"\r\r\n'
```
If I then run the exact same command a second time, I actually receive a SMS and see this in the log:
```
INFO:__main__:sim.command...
Hello User+6421123456"
Result 2 is: b'\r\n> \r\n\r\n+CMGS: 165\r\n\r\nOK\r\n'

```
The SMS when received on my phone looks like this:
```
AT+CMGS="+6421123456"
Hello User
```
I guess I have the format here incorrect? How do I use the number, but only send the actual text message 'Hello User'?
res = await sim_uart.send_command('AT+CMGS="+6421123456"\rHello User\x1A')

Thanks for any help!

**Comment by peterhinch** [comment]
The `SIM_UART` code is ancient and was written for an obsolete version of `uasyncio`. It was also Pyboard specific.

I have updated it to assume an RP2. The `Delay_ms` class is now [in the primitives directory](https://github.com/peterhinch/micropython-async/tree/master/v3/primitives) - you'll need to copy the directory and its contents (or at least `__init__.py` and `delay_ms.py`) to your device.

```python
from machine import UART
import uasyncio as asyncio
from primitives import Delay_ms

class SIM_UART():
    def __init__(self, timeout=4000):
        self.uart = UART(0, 9600, rx=Pin(1, Pin.IN), tx=Pin(0, Pin.OUT))
        self.timeout = timeout
        self.swriter = asyncio.StreamWriter(self.uart, {})
        self.sreader = asyncio.StreamReader(self.uart)
        self.delay = Delay_ms()
        self.response = []
        asyncio.create_task(self.recv())

    async def recv(self):
        while True:
            res = await self.sreader.readline()
            self.response.append(res)  # Append to list of lines
            self.delay.trigger(self.timeout)  # Got something, retrigger timer

    async def send_command(self, command):
        self.response = []  # Discard any pending messages
        await self.swriter.awrite("{}\r\n".format(command))
        print("<", command)
        self.delay.trigger(self.timeout)  # Re-initialise timer
        await self.delay.wait()  # Wait for 4s after last msg received
        return b''.join(self.response)

async def test():
    res = await sim_uart.send_command('rats')
    if res:
        print('Result is:' , res)
    else:
        print('Timed out waiting for result.')

sim_uart = SIM_UART()
try:
    asyncio.run(test())
finally:
    _ = asyncio.new_event_loop()
```

**Comment by andypnz** [comment]
Hi Peter

Thank you. Running your latest code - I needed to make two changes for the SIM7600 - baud rate 115200 and a timeout added to the same line, and I am sending:
res = await sim_uart.send_command('AT+CMGS="+6421123456"\r Hello User\x1A')

Unfortunately the same issue as previously. First time running no SMS is sent and I get:
` Hello User6421123456"
Result is: b'\xffAT+CMGS="+6421123456"\r\r\n'`
Second run a SMS is sent and I get:
` Hello User6421123456"
Result is: b'\xff\r\n> \r\n\r\n+CMGS: 177\r\n\r\nOK\r\n'`
The received SMS on my phone looks like this:
`AT+CMGS="+6421123456"
 Hello User`
How can I send a SMS on the first run? I guess I have the format here incorrect? How do I use the number, but only send the actual text message 'Hello User'?
Thank you again for any help
Andrew

**Reply by 2dof** [reply]
Before sending SMS message , put some commands for status checking  like: , sim status registration checking/ enabling  status; SMS message config  etc.

Im not sure (it is my interpretation), but it look like at the first attempt of sending sms modem accept only  phone number, and with the second attempt rest of  command ? 
Try send message commands separately with some small delay between (like from arduino examples)  . ie  
commands 
``` 
 AT+CMGF=1 +<enter>                            # Set to text mode
AT+CMGS="\" + <number> + "\" <enter>   # Set number to send SMS
<message> +<CTRL-Z>                                 # txt message#  CTRL+Z  (= hex: 1A )to end SMS  and send 
                                                  
```
(send_command() has all ready  <enter> ( \r) , but try modify and add <enter>("\r" )  and <CTRL-Z>( "1A" ) manually in commands.
go to  section [Sending the message](https://www.smssolutions.net/tutorials/gsm/sendsmsat/) for details.

**Reply by peterhinch** [reply]
@andypnz I can help with `uasyncio` but I have no experience with SIM7600. Sorry.

My general approach to this type of problem would be to connect a terminal program to the device and learn how to drive it manually. Then modify the code to match.

**Comment by andypnz** [comment]
Thank you @peterhinch @2dof (sorry about the formatting below! Trying to sort out)
Splitting the number + message solved both the messy received SMS and also the not sending on the first run issue :-) I don't have much programming knowledge - so my method to send the two serial messages was to create a main file + class.... and then in my main file create the two messages and send using the class.

My problem now is although it correctly sends a SMS, it crashes with the error below. I think an issue in     async def recv(self) ?
```sms-loop is running
< AT+CMGS="+6421123456"
Result 1 is: b'\xffAT+CMGS="+6421123456"\r\r\n'
< Hello User
Timed out waiting for result.
here
sms-loop is running
here
sms-loop is running
here
Task exception wasn't retrieved
future: <Task> coro= <generator object 'recv' at 2000afc0>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "class_sim_800.py", line 18, in recv
  File "uasyncio/stream.py", line 1, in readline
TypeError: unsupported types for __iadd__: 'bytes', 'NoneType'
Traceback (most recent call last):
  File "<stdin>", line 38, in <module>
  File "uasyncio/core.py", line 1, in run_forever
  File "uasyncio/core.py", line 1, in run_until_complete
  File "class_sim_800.py", line 18, in recv
  File "uasyncio/stream.py", line 1, in readline
KeyboardInterrupt: 
>>> ```

main.py:
```import uasyncio as asyncio
from machine import UART, Pin
from class_sim_800 import SIM_UART
first_time_sms = 0
async def sms_loop():
  global first_time_sms  
  while True:  
    await asyncio.sleep(1) # Delay at begin
    print("sms-loop is running")
    sim_uart = SIM_UART() 
    if (first_time_sms == 0):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(1)
        res = await sim_uart.send_command('AT+CMGS="+6421123456"\r')
        if res:
            print('Result 1 is:' , res)
        else:
            print('Timed out waiting for result.')
    if (first_time_sms == 1):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(1)
        res = await sim_uart.send_command('Hello User\x1A')
        if res:
            print('Result 2 is:' , res)
        else:
            print('Timed out waiting for result.')         
    if (first_time_sms > 1):
        print('here')
loop = asyncio.get_event_loop()
f=asyncio.create_task(sms_loop())
loop.run_forever()```


class_sim_800.py
```from machine import UART, Pin
import uasyncio as asyncio
from primitives import Delay_ms
class SIM_UART():
    def __init__(self, timeout=4000):
        self.uart = UART(0, 115200, rx=Pin(1, Pin.IN), tx=Pin(0, Pin.OUT), timeout=4000)
        self.timeout = timeout
        self.loop = asyncio.get_event_loop()
        self.swriter = asyncio.StreamWriter(self.uart, {})
        self.sreader = asyncio.StreamReader(self.uart)
        self.delay = Delay_ms()
        self.response = []
        asyncio.create_task(self.recv())

    async def recv(self):
        while True:
            res = await self.sreader.readline()
            self.response.append(res)  # Append to list of lines
            self.delay.trigger(self.timeout)  # Got something, retrigger timer

    async def send_command(self, command):
        self.response = []  # Discard any pending messages
        await self.swriter.awrite("{}\r\n".format(command))
        print("<", command)
        self.delay.trigger(self.timeout)  # Re-initialise timer
        await self.delay.wait()  # Wait for 4s after last msg received
        return b''.join(self.response)```

**Reply by peterhinch** [reply]
I think the problem here is that the UART is timing out. Try setting the timeouts to a very long value:
```py
self.uart = UART(0, 115200, rx=Pin(1, Pin.IN), tx=Pin(0, Pin.OUT), timeout=1000_000, timeout_char=1000_000)
```

**Comment by andypnz** [comment]
Hi Peter
I tried as you suggested (and also by another factor of 10) - it delays the 'Timed out waiting for result.' and 'Task exception wasn't retrieved' errors - but the errors remain the same.....
More than happy to try anything else you may suggest.... (also updated to the latest RP2040 nightly - no change...)

**Reply by 2dof** [reply]
does error occur when You send sms like I suggested (i.e) put commands, message separately ?  if yes , note that between sending command. message delay should be small (for example 100 ms ). when send module should return (after 4-5 sec) some status/confirmation code  in format +CMGS:<mr> . mr : number .
but in your code in  in send command you have timeout set for 4 sec (4000 ms ) and after that in main code ```await asyncio.sleep(1)`` 
try to adapt code  for 100ms waitng, and then 4 -5 sec for waintg for responce. 

  code error ``` TypeError: unsupported types for __iadd__: 'bytes', 'NoneType' ``` 
looks like  [rp2: Occasional UART failures #8344](https://github.com/micropython/micropython/issues/8344)   issue - I may be wrong so maybe Peter  will help.

**Reply by peterhinch** [reply]
Iss #8344 was fixed.

**Reply by andypnz** [reply]
@2dof Thank you for the suggestion - I have tried various delays with no change in behavior....

**Comment by peterhinch** [comment]
@andypnz I'm not clear what you're seeing. With my suggested value of 1,000,000 the timeout should be 1,000s or about 16.6 minutes. Please let me know how long it actually takes to fail.

FYI I raised https://github.com/micropython/micropython/issues/10867 in February as UART timeouts keep causing `uasyncio.StreamReader` to break. In my opinion it should be possible to set the UART timeouts to infinity - `uasyncio` has its own timeout mechanism.

**Comment by andypnz** [comment]
@peterhinch With the 1,000,000 timeout, it delays the 'Timed out waiting for result.' and 'Task exception wasn't retrieved' errors by maybe 10 seconds.

With the 1,000,000 timeout, and my code changed to send 2 x SMS's:
```async def sms_loop():
  global first_time_sms  
  while True:  
    #await asyncio.sleep(1) # Delay at begin
    print("sms-loop is running")
    sim_uart = SIM_UART() 
    if (first_time_sms == 0):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(0.1)
        res = await sim_uart.send_command('AT+CMGS="+6421123456"\r')
        if res:
            print('Result 0 is:' , res)
        else:
            print('Timed out waiting for result 0.')
    if (first_time_sms == 1):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(0.1)
        res = await sim_uart.send_command('Hello User\x1A')
        if res:
            print('Result 1 is:' , res)
        else:
            print('Timed out waiting for result 1.')
    if (first_time_sms == 2):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(4)
        res = await sim_uart.send_command('AT+CMGS="+6421123456"\r')
        if res:
            print('Result 2 is:' , res)
        else:
            print('Timed out waiting for result 2.')
    if (first_time_sms == 3):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(0.1)
        res = await sim_uart.send_command('Hello UserTHREE\x1A')
        if res:
            print('Result 3 is:' , res)
        else:
            print('Timed out waiting for result 3.')                           
    if (first_time_sms > 3):
        print('here')```

It correctly sends the two SMS messages (Hello User and Hello UserTHREE are received correctly on my phone). But I still see these timeouts:

```sms-loop is running
< AT+CMGS="+6421123456"
Result 0 is: b'\xffAT+CMGS="+6421123456"\r\r\n'
< Hello User
Timed out waiting for result 1.
< AT+CMGS="+6421123456"
Timed out waiting for result 2.
< Hello UserTHREE
Result 3 is: b'+SMS FULL\r\n'
here
sms-loop is running
here
sms-loop is running
here
sms-loop is running
here```

And then after around 2 minutes, I get this continuous list of errors:

```sms-loop is running
here
sms-loop is running
here
sms-loop is running
Task exception wasn't retrieved
future: <Task> coro= <generator object 'sms_loop' at 2000ae30>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 13, in sms_loop
  File "class_sim_800.py", line 13, in __init__
  File "primitives/delay_ms.py", line 28, in __init__
MemoryError: memory allocation failed, allocating %u bytes
Task exception wasn't retrieved
future: <Task> coro= <generator object 'recv' at 20011cc0>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "class_sim_800.py", line 19, in recv
MemoryError: memory allocation failed, allocating 52 bytes
Task exception wasn't retrieved
future: <Task> coro= <generator object '_run' at 20012230>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/core.py", line 1, in queue_read
  File "uasyncio/core.py", line 1, in _enqueue
MemoryError: memory allocation failed, allocating %u bytes
Task exception wasn't retrieved
future: <Task> coro= <generator object 'recv' at 200121a0>
Traceback (most recent call last):```

**Reply by andypnz** [reply]
I also tried using a ESP32 - 2 x SMS send with the same timeout issue, but the big error list happens much sooner:
```sms-loop is running
here
sms-loop is running
Task exception wasn't retrieved
future: <Task> coro= <generator object 'sms_loop' at 3ffe5d40>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 13, in sms_loop
  File "class_sim_800.py", line 13, in __init__
MemoryError: memory allocation failed, allocating 96 bytes
Task exception wasn't retrieved
future: <Task> coro= <generator object 'recv' at 3ffea880>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/stream.py", line 1, in readline
  File "uasyncio/core.py", line 1, in queue_read
MemoryError: ```
(note the error list continues ?indefinatly?)

**Comment by andypnz** [comment]
Taking a step backwards - it I only send 'AT' as the command, this is the result. Each command returns a result, but with errors between - except the first which returns with no error.

```>>> %Run -c $EDITOR_CONTENT
sms-loop is running
< AT

Result -1 is: b'AT\r\r\nOK\r\n'
sms-loop is running
< AT

Task exception wasn't retrieved
future: <Task> coro= <generator object 'recv' at 3ffe64f0>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "class_sim_800.py", line 19, in recv
  File "uasyncio/stream.py", line 1, in readline
TypeError: unsupported types for __iadd__: 'bytes', 'NoneType'
Result -1 is: b'OK\r\n'
sms-loop is running
< AT

Task exception wasn't retrieved
future: <Task> coro= <generator object 'recv' at 3ffe9070>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "class_sim_800.py", line 19, in recv
  File "uasyncio/stream.py", line 1, in readline
TypeError: unsupported types for __iadd__: 'bytes', 'NoneType'
Result -1 is: b'OK\r\n'
sms-loop is running```

**Comment by peterhinch** [comment]
Sadly the Pico UART isn't working properly, and is timing out far too early. This can be proved by pasting the following (link pins 0 and 1):
```python
from machine import UART, Pin
uart = UART(0, 115200, rx=Pin(1, Pin.IN), tx=Pin(0, Pin.OUT), timeout=1000_000, timeout_char=1000_000)
uart.readline()
```
Times out after 15s rather than 16.6mins. I have raised an issue https://github.com/micropython/micropython/issues/11678.

**Reply by andypnz** [reply]
I tried the same test on a ESP32 and a Nucleo L476RG - with the same result - times out after around 15s.....

I guess I need to await this issue to be solved prior to trying to debug this? Please let me know if you hear anything :-)

**Reply by peterhinch** [reply]
I think the best that can be done is to do
```python
uart = UART(0, 115200, rx=Pin(1, Pin.IN), tx=Pin(0, Pin.OUT), timeout=65535, timeout_char=65535)
```
which gives a timeout of 1min 5.5s. Then create a task which sends an AT command every (say) 50s. Clearly you need to send something harmless that merely triggers the modem into sending a response and stopping the receiver from timing out. The receive loop then recognises and discards the "harmless" response.

**Reply by andypnz** [reply]
Hi Peter. Unfortunately I don't think I have implemented your suggestion correctly. This code does not send any SMS, and errors as below:

```import uasyncio as asyncio
from machine import UART, Pin
from class_sim_800 import SIM_UART
first_time_sms = 0
async def at_sms_loop(): 
  while True:  
    print("AT sms-loop is running")
    await asyncio.sleep(50) # Delay at begin
    sim_uart = SIM_UART() 
    res = await sim_uart.send_command('AT\r')
    if res:
        print('AT loop result is:' , res)
    else:
        print('Timed out waiting for AT result.') 

async def sms_loop():
  global first_time_sms  
  while True:  
    #await asyncio.sleep(1) # Delay at begin
    print("sms-loop is running")
    sim_uart = SIM_UART() 
    if (first_time_sms == 0):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(0.1)
        res = await sim_uart.send_command('AT+CMGS="+6421976665"\r')
        if res:
            print('Result 0 is:' , res)
        else:
            print('Timed out waiting for result 0.')
    if (first_time_sms == 1):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(0.1)
        res = await sim_uart.send_command('Hello User\x1A')
        if res:
            print('Result 1 is:' , res)
        else:
            print('Timed out waiting for result 1.')
            
    if (first_time_sms == 2):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(4)
        res = await sim_uart.send_command('AT+CMGS="+6421976665"\r')
        if res:
            print('Result 2 is:' , res)
        else:
            print('Timed out waiting for result 2.')
    if (first_time_sms == 3):
        first_time_sms = first_time_sms + 1
        await asyncio.sleep(0.1)
        res = await sim_uart.send_command('Hello UserTHREE\x1A')
        if res:
            print('Result 3 is:' , res)
        else:
            print('Timed out waiting for result 3.')             
                   
    if (first_time_sms > 3):
        print('here')
        
loop = asyncio.get_event_loop()
e=asyncio.create_task(at_sms_loop())
f=asyncio.create_task(sms_loop())
loop.run_forever()
```

class_sim_800.py:
```from primitives import Delay_ms

class SIM_UART():
    def __init__(self, timeout=4000):
        self.uart = UART(0, 115200, rx=Pin(1, Pin.IN), tx=Pin(0, Pin.OUT), timeout=65535, timeout_char=65535)
        self.timeout = timeout
        self.loop = asyncio.get_event_loop()
        self.swriter = asyncio.StreamWriter(self.uart, {})
        self.sreader = asyncio.StreamReader(self.uart)
        self.delay = Delay_ms()
        self.response = []
        asyncio.create_task(self.recv())

    async def recv(self):
        while True:
            res = await self.sreader.readline()
            self.response.append(res)  # Append to list of lines
            self.delay.trigger(self.timeout)  # Got something, retrigger timer

    async def send_command(self, command):
        self.response = []  # Discard any pending messages
        await self.swriter.awrite("{}\r\n".format(command))
        print("<", command)
        self.delay.trigger(self.timeout)  # Re-initialise timer
        await self.delay.wait()  # Wait for 4s after last msg received
        return b''.join(self.response)
```

```AT sms-loop is running
sms-loop is running
< AT+CMGS="+6421976665"
< AT
Task exception wasn't retrieved
future: <Task> coro= <generator object 'recv' at 2000e810>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "class_sim_800.py", line 19, in recv
  File "uasyncio/stream.py", line 1, in readline
TypeError: unsupported types for __add__: 'bytes', 'NoneType'
Timed out waiting for AT result.
AT sms-loop is running
Result 0 is: b'\xffAT+CMGS=AT\r\r\nERROR\r\n'
< Hello User
Timed out waiting for result 1.
< AT+CMGS="+6421976665"
Timed out waiting for result 2.
< Hello UserTHREE
Timed out waiting for result 3.
here
sms-loop is running
here
sms-loop is running
here
sms-loop is running
....
....
sms-loop is running
here
sms-loop is running
< AT
Task exception wasn't retrieved
future: <Task> coro= <generator object 'sms_loop' at 2000cb30>
Traceback (most recent call last):
  File "uasyncio/core.py", line 1, in run_until_complete
  File "<stdin>", line 24, in sms_loop
  File "class_sim_800.py", line 13, in __init__
  File "primitives/delay_ms.py", line 30, in __init__
MemoryError: memory allocation failed, allocating 96 bytes
```
and a list of similar errors..........`

**Reply by andypnz** [reply]
If I leave out the at_sms_loop, I get sent the 2 x SMS's, but then eventually crashes as we know

```
sms-loop is running
< AT+CMGS="+6421976665"
Result 0 is: b'+SMS FULL\r\n\r\n> \r\n'
< Hello User
Timed out waiting for result 1.
< AT+CMGS="+6421976665"
Timed out waiting for result 2.
< Hello UserTHREE
Timed out waiting for result 3.
here
sms-loop is running
here
sms-loop is running
here
sms-loop is running
here
```

**Comment by robert-hh** [comment]
It seems that the timeout variable is 16 bit in each and every port I looked into. So the largest reasonable value is 65535 at the moment.

**Comment by andypnz** [comment]
@peterhinch Peter - Do you think you would have time to try and solve this for the community if I sent you a SIM7600 (to keep)? I don't have the skills here, and I do think this would be a useful library to have (and be working with uasyncio)?

**Reply by peterhinch** [reply, marked as answer]
Thanks for the offer but I'm not in a position to take this on as a project, but I will help with the `uasyncio` and UART code. You'll need to research the AT codes.

The following makes use of the timeout feature which is new to `uasyncio` V3. I have tested it with a loopback. It solves the UART timeout problem by periodically (40 sec intervals) sending "AT" to the modem. As I understand it, this will cause it to respond and will keep the UART receiver alive. The `send_command` method waits 4s to gather any response from the modem.
```python
from machine import UART, Pin
import uasyncio as asyncio
from primitives import Delay_ms

class SIM_UART():
    def __init__(self):
        self.uart = UART(0, 115200, rx=Pin(1, Pin.IN), tx=Pin(0, Pin.OUT), timeout=65_535, timeout_char=65_535)
        while self.uart.any():
            self.uart.read(1)  # Clear any initial garbage
        self.swriter = asyncio.StreamWriter(self.uart, {})
        self.sreader = asyncio.StreamReader(self.uart)
        self.delay = Delay_ms(duration=40_000)  # 40s (less than UART timeouts)
        self.lock = asyncio.Lock()
        asyncio.create_task(self.keep_alive())

    async def keep_alive(self):
        while True:
            await self.delay.wait()
            self.delay.clear()
            res = await self.send_command("AT")
            if not res:
                raise OSError("Modem not responding")
            
    async def send_command(self, command):
        res = b""
        async with self.lock:  # Don't allow concurrent messages
            await self.swriter.awrite(f"{command}\r\n")
            done = False
            while not done:
                try:
                    r = await asyncio.wait_for(self.sreader.readline(), 4)
                    self.delay.trigger()
                    res = b"".join((res, r))
                except asyncio.TimeoutError:
                    done = True  # No more data from device
        return res


async def test():
    res = await sim_uart.send_command('AT')
    if res:
        print('Result is:' , res)
    else:
        print('Timed out waiting for result.')
    await asyncio.sleep(120)  # Test keep_alive

sim_uart = SIM_UART()
try:
    asyncio.run(test())
finally:
    _ = asyncio.new_event_loop()

# await sim_uart.send_command("AT+CMGS=13012832788")  # Phone no.
# await sim_uart.send_command("hello\x1B")  # Terminate with ESC (0x1B) or ctrl-z (0x1A)
```
The final comments are my guess at what is needed to send an SMS. The manual indicates that the text must be terminated with `ESC` or `ctrl-z`. This allows for multi-line messages.

**Reply by andypnz** [reply]
Hi @peterhinch Peter 
Thank you very much for your time! That code solves all the timeout / error issues! I will now (as you suggest) work on sorting the actual AT commands. Currently the SMS received on my phone is populated with many leading "AT" statements. It is like there is a send buffer which needs to be cleared - which I don't understand as the message to send is supposed to come after the CMGS and phone number command..... I will research this and come back here to complete the solution. Thanks again!

**Reply by peterhinch** [reply]
The `keep_alive` method sends `AT\r\n`. I assumed that the modem would just return with a prompt. There are other options: all that's required is a command which prompts the modem to respond. If this command is misbehaving and finding its way into texts, there are other commands which return the serial number or modem type which could be used instead.

My memory of driving modems with AT strings is flaky: I last did it in the days of dial-up internet.

**Reply by andypnz** [reply]
Hi Peter

Ok, I think the problem is the keep alive AT command is not synchronized with sending a message. So the keep alive AT command  ends up in the message string - and appears on my phone. There is no AT command possible that does not have a response of some sort.

```test loop running
Result -1 is: b'AT+CMGF=1\r\r\nOK\r\n'
Result 0 is: b'AT+CMGS="+6421123456"\r\r\n'
keep alive result is: b'\r\n'
Result 1 is: b'\r\n\r\n+CMGS: 3\r\n\r\nOK\r\n'
keep alive result is: b'AT\r\r\nOK\r\n'
Result 2 is: b'AT+CMGS="+6421123456"\r\r\n'
keep alive result is: b'\r\n'
Result 3 is: b'\r\n\r\n+CMGS: 4\r\n\r\nOK\r\n'
here
keep alive result is: b'AT\r\r\nOK\r\n'
keep alive result is: b'AT\r\r\nOK\r\n'
keep alive result is: b'AT\r\r\nOK\r\n'
>>> 
```

If I remove the keep alive AT command from the class, and instead send some 'keep alive' AT commands within async def test():  (and therefore coordinated with sending an actual SMS) all works as expected. It is almost like we need a flag to only send a keep alive AT command if there is no SMS about to be sent? Is that what you would think? What would be the best implementation?

And finally to receive a SMS from my phone - I guess the thing to do is to keep checking the modem for new messages on a regular basis (say every 0.5 sec) using the get unread message AT command? 
``` 'AT+CMGL="REC UNREAD"\r\n'       #gets incoming SMS that has not been read  ```
Is this best within the class, or async def test():? 
Actually, if this is what needs to happen, it would keep the UART alive :-)

Thank you!
Andrew

**Reply by peterhinch** [reply]
If you are going to regularly poll the device for incoming SMS that would indeed keep the UART receiver alive. You would still need synchronisation to ensure that your polling did not occur while you were sending an SMS, otherwise you would get the same problem as you have now.

Evidently my synchronisation isn't working. I'll explain how it's supposed to work, and maybe one of us will spot the flaw. Firstly the `Lock` ensures that two transmissions to the modem can't occur concurrently. The `send_command` method must run to completion before another can start. Completion includes receiving data from the UART and waiting until 4s has elapsed after the last line was received.

Reception of a line retriggers the `Delay_ms` object. This will only time out when nothing has been received for 40s, when it will send `AT\r\n`. Because this uses the `send_command` method, the latter's `Lock` will prevent an SMS from being sent until 4s after the modem has responded to the `AT\r\n`.

The 4s timeout does assume that the modem responds to a command within 4s and then remains silent until it gets another command. If it sends unsolicited messages or pauses for >4s between lines, then the extra data will be stored in the modem receive buffer causing the next `send_command` to return the extra data. If the modem sends unsolicited data the design needs to change. If its response is slow, the 4s timeout needs to be increased,

**Comment by peterhinch** [comment]
One thing you could try is to amend the `send_command` method thus:
```python
    async def send_command(self, command):
        res = b""
        uns = b""
        async with self.lock:  # Don't allow concurrent messages
            # Collect and print any unsolicited messages
            done = False
            while not done:
                try:
                    r = await asyncio.wait_for(self.sreader.readline(), 1)
                    uns = b"".join((uns, r))
                except asyncio.TimeoutError:
                    done = True  # No more data from device
            if uns:
                print(f"Unsolicited: {uns}")
            # UART buffer is now empty: send the command and await the response
            await self.swriter.awrite(f"{command}\r\n")
            done = False
            while not done:
                try:
                    r = await asyncio.wait_for(self.sreader.readline(), 4)
                    self.delay.trigger()
                    res = b"".join((res, r))
                except asyncio.TimeoutError:
                    done = True  # No more data from device
        return res
```
This will report any unsolicited messages or responses to the prior message that occurred after the 4s timeout had elapsed.

**Reply by andypnz** [reply]
Hi Peter

Thanks again! I will look at this more in the light of day tomorrow....
But I have managed to get:
```
Unsolicited: b'\r\n+CMTI: "SM",0\r\n'
```
This occurred with your new send_command (but without using your keep alive loop) and me sending a receive unread request every 10 seconds with flags so nothing gets sent more frequently than every 10 seconds. I will also try with your full code tomorrow.

Overall everything is working (I just need to parse out the message bit of the received message) - just this (I think unexpected) 'unsolicited'.

Hope you are having a good weekend
Andrew

**Reply by andypnz** [reply]
Hi Peter

Unfortunately no real progress. Try as I might I am unable to give a concise analysis of the cause for the 'unsolicited' responses.  

Fortunately these don't seem to have an impact on the functionality - which is working as expected. Rather than parsing the messages, I am using string.find() as I am just looking for a command to execute. And I use the same technique to check the phone number the message is from matches my white list. And confirming the phone number is present in the message means I only note an unknown command which is real.

When I have some time I may put an analyser on the data stream and see if I can figure out a bit more. 

Thanks again for all your help :-)
